sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/MergeRowsExec.scala

```diff
@@ -0,0 +1,218 @@
+
+    val matchedInstructionExecs = planInstructions(matchedInstructions)
+    val notMatchedInstructionExecs = planInstructions(notMatchedInstructions)
+    val notMatchedBySourceInstructionExecs = planInstructions(notMatchedBySourceInstructions)
+
+    val cardinalityValidator = if (checkCardinality) {
+      val rowIdOrdinal = child.output.indexWhere(attr => conf.resolver(attr.name, ROW_ID))
+      assert(rowIdOrdinal != -1, "Cannot find row ID attr")
+      BitmapCardinalityValidator(rowIdOrdinal)
+    } else {
+      NoopCardinalityValidator
+    }
+
+    val mergeIterator = new MergeRowIterator(
+      rowIterator, cardinalityValidator, isTargetRowPresentPred, isSourceRowPresentPred,
+      matchedInstructionExecs, notMatchedInstructionExecs, notMatchedBySourceInstructionExecs)
+
+    // null indicates a record must be discarded
+    mergeIterator.filter(_ != null)
+  }
+
+  private def createProjection(exprs: Seq[Expression]): UnsafeProjection = {
+    UnsafeProjection.create(exprs, child.output)
+  }
+
+  private def createPredicate(expr: Expression): BasePredicate = {
+    GeneratePredicate.generate(expr, child.output)
+  }
+
+  private def planInstructions(instructions: Seq[Instruction]): Seq[InstructionExec] = {
+    instructions.map {
+      case Keep(cond, output) =>
+        KeepExec(createPredicate(cond), createProjection(output))
+
+      case Split(cond, output, otherOutput) =>
+        SplitExec(createPredicate(cond), createProjection(output), createProjection(otherOutput))
+
+      case other =>
+        throw new AnalysisException(s"Unexpected instruction: $other")
+    }
+  }
+
+  sealed trait InstructionExec {
+    def condition: BasePredicate
+  }
+
+  case class KeepExec(condition: BasePredicate, projection: Projection) extends InstructionExec {
+    def apply(row: InternalRow): InternalRow = projection.apply(row)
+  }
+
+  case class SplitExec(
+      condition: BasePredicate,
+      projection: Projection,
+      otherProjection: Projection) extends InstructionExec {
+    def projectRow(row: InternalRow): InternalRow = projection.apply(row)
+    def projectExtraRow(row: InternalRow): InternalRow = otherProjection.apply(row)
+  }
+
+  sealed trait CardinalityValidator {
+    def validate(row: InternalRow): Unit
+  }
+
+  object NoopCardinalityValidator extends CardinalityValidator {
+    def validate(row: InternalRow): Unit = {}
+  }
+
+  /**
+   * A simple cardinality validator that keeps track of seen row IDs in a roaring bitmap.
+   * This validator assumes the target table is never broadcasted or replicated, which guarantees
+   * matches for one target row are always co-located in the same partition.
+   *
+   * IDs are generated by [[org.apache.spark.sql.catalyst.expressions.MonotonicallyIncreasingID]].
+   */
+  case class BitmapCardinalityValidator(rowIdOrdinal: Int) extends CardinalityValidator {
+    private val matchedRowIds = new Roaring64Bitmap()
```

Comment:
Could you add some description why we need to choose `Roaring64Bitmap` instead of `RoaringBitmap`? I'm interested in the background of decision in this use case.

Reply:
Added above.